include/seqan3/range/view/pairwise_combine.hpp

```diff
@@ -0,0 +1,595 @@
+                             ((src_size - index_i) - 1)/2;
+            first_it = begin_it + index_i;
+            second_it = begin_it + index_j;
+        }
+
+        //!\brief The iterator pointing to the first element of the pairwise combination.
+        source_iterator_type first_it{};
+        //!\brief The iterator pointing to the second element of the pairwise combination.
+        source_iterator_type second_it{};
+        //!\brief The begin of the source range.
+        source_iterator_type begin_it{};
+        //!\brief The end of the source range.
+        source_iterator_type end_it{};
+    };
+
+public:
+
+    /*!\name Constructors, destructor and assignment
+     * \{
+     */
+    constexpr pairwise_combine_view() = default;
+    constexpr pairwise_combine_view(pairwise_combine_view const &) = default;
+    constexpr pairwise_combine_view(pairwise_combine_view &&) = default;
+    constexpr pairwise_combine_view & operator=(pairwise_combine_view const &) = default;
+    constexpr pairwise_combine_view & operator=(pairwise_combine_view &&) = default;
+    ~pairwise_combine_view() = default;
+
+    /*!\brief Constructs from a view.
+     * \param[in] range A rvalue reference to the source view.
+     *
+     * \details
+     *
+     * During construction the iterator pointing to the last element of the view is cached. This optimises
+     * the call to end if the source range models only ForwardRange. Otherwise the call to end will be
+     * linear in the number of elements of the source range.
+     */
+    explicit constexpr pairwise_combine_view(source_range_type && range) : src_range{std::move(range)}
+    {
+        // Check if range is empty.
+        if (std::ranges::begin(src_range) == std::ranges::end(src_range))
+        {
+            back_iterator = std::ranges::end(src_range);
+        }
+        else
+        {
+            if constexpr (std::ranges::BidirectionalRange<source_range_type> &&
+                          std::ranges::CommonRange<source_range_type>)
+            {
+                back_iterator = std::ranges::prev(std::ranges::end(src_range));
+            }
+            else
+            {  // Cannot decrement so we invoke linear pass until we get to the last element.
+                size_t size = std::distance(src_range);
+                back_iterator = std::ranges::begin(src_range);
+                std::advance(back_iterator, size - 1);
+            }
+        }
+    }
+
+    /*!\brief Constructs from a range.
+     * \param[in] range A lvalue reference to the source range.
+     *
+     * \details
+     *
+     * Wraps the passed range in a ref_view in order to store lvalue references to ranges (non-views) as a view
+     * and delegate to the view constructor above.
+     */
+    template <typename other_range_type>
+    explicit constexpr pairwise_combine_view(other_range_type & range) : pairwise_combine_view{view::all(range)}
```

Comment:
Well explicit is a habit. They are not constrained because if they are it breaks a lot of stuff since it won't be move constructible anymore. I don't have the time to investigate this so we need to live with this for the moment.

Reply:
Also note, that this constructor was added because of the C++20 standard. But AFAIK these are not implemented in the current range versions (at least random lookups didn't reveal any hint on how/where they might be implemented. Don't know if this is tackled in future versions of the ranges library or if you know more about, then please enlighten me. But I can also just leave this ctor away because it just gives me PITA that I don't have time for.

Reply:
Ok, I did even more digging and the strangest things happen. I did find the biggest problem and resolved it, but there are still some open investigations that need to be done. See the todo comment that I left in the code.